### DISCUSSION TITLE
Minio console behind secured NGINX reverse proxy

### DISCUSSION THREAD

**Discussion by FeryET** [minio/minio#14819] [Q&A]
I am trying to make minio console accessible at `domain.com/minio` using nginx reverse proxy. You can see a sample repository here: https://github.com/FeryET/minio-nginx-docker-test

The problem I am having here right now is that, for whatever reason, loading MinIO behind a https reverse proxy using nginx breaks minio. You can see this here: https://farhoodetaati.me/minio. If you curl this address (with `-k` flag), you will see there is this line in the xml:
```XML
<noscript>You need to enable JavaScript to run this app.</noscript>
```

I have also tried sharing the certification between MinIO and nginx, but this does not alleviate the issue. The thing is, if you change the nginx image from `nginx-certbot` to `nginx` and disable SSL connection, listen to port 80 using `nginx.conf` in the repo, the website clearly works. There's something fishy with how SSL connection is being handled by MinIO console.

**Comment by harshavardhana** [comment]
you need to set MINIO_BROWSER_REDIRECT_URL=http://domain.com/minio

**Reply by FeryET** [reply]
I've also set that. Not helping. Does not alleviate the issue.

**Reply by harshavardhana** [reply]
you need to use latest release of MinIO

**Reply by FeryET** [reply]
I'm using the latest docker image of MinIO (image id: dc8ab0b84bb9). If you look at https://farhoodetaati.me/minio the problem still exists.

**Comment by Monstermash28425R1** [comment]
i have some issue too deploying my minio instance behind a nginx proxy, can anybody help me? i served my minio console successfuly but when i login and try to browse any bucket it show the loading message but never load anything, but when i open my console instance with the server-ip:port y can browse all my buckets, im serving my console with a ssl configuration inside nginx


this is my nginx config

        server {
                server_name browser.io.example.com;
                #ignore_invalid_headers off;
                client_max_body_size 100m;
                proxy_buffering off;
        
                location / {
                        proxy_set_header Host $host;
                        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                        proxy_set_header X-Real-IP $remote_addr;
                        proxy_set_header X-Scheme $scheme;
                        proxy_set_header X-Forwarded-Proto $scheme;
        
                        proxy_redirect off;
                        proxy_connect_timeout 300;
                        proxy_http_version 1.1;
        
                        #proxy_set_header Connection "";
                        #chunked_transfer_encoding off;
                        proxy_pass http://localhost:9001/;
                }
        
             # ssl config goes here
        }


should i open a discussion for this?

**Reply by harshavardhana** [reply]
You need to ensure websocket works for the console proxy upstream via supporting connection upgrade.

**Reply by AlexZIX** [reply]
Try to add headers for support the WebSocket connections via proxy. We use this:

# WebSocket support (nginx 1.4)
proxy_http_version			1.1;
proxy_set_header			Upgrade				$http_upgrade;
proxy_set_header			Connection			$connection_upgrade;

**Comment by jabbors** [comment]
I attempted to try proxying console requests behind HAProxy to see if it works differently.

I began with starting MinIO without configuring MINIO_BROWSER_REDIRECT_URL followed by proxying all traffic entering HAProxy to a backend for the console. This worked fine. The login page loaded, I could login and operate MinIO.

Then I restarted MinIO, but this time I configured MINIO_BROWSER_REDIRECT_URL=https://my.domain.com/console. Now the login page failed to load and I get the same response `You need to enable JavaScript to run this app`.

Note: All requests where made over SSL where HAProxy takes care of SSL termination.

**Comment by cinhtau** [comment]
I managed to setup my Nginx properly, the site is loading properly

I set `MINIO_BROWSER_REDIRECT_URL` as well and it did not work.

```
You need to enable JavaScript to run this app.
```

the access to the resources is fine, as all css and js files could be loaded.

```html
<html>

<body>
    <!--StartFragment-->

    |
    <!doctype html>
    <html lang="en">

    <head>
        <meta charset="utf-8" />
        <base href="/minio/" />
        <meta content="width=device-width,initial-scale=1" name="viewport" />
        <meta content="#081C42" media="(prefers-color-scheme: light)" name="theme-color" />
        <meta content="#081C42" media="(prefers-color-scheme: dark)" name="theme-color" />
        <meta content="MinIO Console" name="description" />
        <meta name="minio-license" content="agpl" />
        <link href="./styles/root-styles.css" rel="stylesheet" />
        <link href="./apple-icon-180x180.png" rel="apple-touch-icon" sizes="180x180" />
        <link href="./favicon-32x32.png" rel="icon" sizes="32x32" type="image/png" />
        <link href="./favicon-96x96.png" rel="icon" sizes="96x96" type="image/png" />
        <link href="./favicon-16x16.png" rel="icon" sizes="16x16" type="image/png" />
        <link href="./manifest.json" rel="manifest" />
        <link color="#3a4e54" href="./safari-pinned-tab.svg" rel="mask-icon" />
        <title>MinIO Console</title>
        <script defer="defer" src="./static/js/main.d1382a00.js"></script>
        <link href="./static/css/main.57e739f5.css" rel="stylesheet">
    </head>

    <body><noscript>You need to enable JavaScript to run this app.</noscript>
        <div id="root">
            <div id="preload"><img src="./images/background.svg" /> <img src="./images/background-wave-orig2.svg" /></div>
            <div id="loader-block"><img src="./Loader.svg" /></div>
        </div>
    </body>

    </html>
    -- | --

    <!--EndFragment-->
</body>

</html>
```

Inspecting the js in `useApi.tx` error, calling the url `api/v1/admin/site-replication` does not work.

```javascript
const useApi = (
  onSuccess: NoReturnFunction,
  onError: NoReturnFunction
): [IsApiInProgress, ApiMethodToInvoke] => {
  const [isLoading, setIsLoading] = useState<boolean>(false);

  const callApi = (method: string, url: string, data?: any, headers?: any) => {
    setIsLoading(true);
    api
      .invoke(method, url, data, headers)
      .then((res: any) => {
        setIsLoading(false);
        onSuccess(res);
      })
      .catch((err: ErrorResponseHandler) => {
        setIsLoading(false);
        onError(err);
      });
  };

  return [isLoading, callApi];
};
```

The console is not usable though, due to this error.

* MINIO_SERVER_URL &rArr; API
* MINIO_BROWSER_REDIRECT_URL &rArr; Console

We stop using nginx as reverse proxy and use `--certs` option to serve TLS directly.

**Comment by washanhanzi** [comment]
I don't think minio work with reverse proxy with subpath. I must use a dedicate domain name to proxy minio console.

https://mydomain.com/minio -> https://minio.mydomain.com/
